feat(TwabRewards): add endPromotion function

PierrickGT opened the pull request:
Issue: https://github.com/code-423n4/2021-12-pooltogether-findings/issues/75

asselstine reviewed (changes requested):
`destroyPromotion` should play nicely with the existing `cancelPromotion`.

To better define them, and to ensure they have good interplay, let's do this:

- Rename `cancelPromotion` to `endPromotion`: because it allows the creator to prematurely end the promotion.  
- When the creator calls `endPromotion`, the code actually adjusts the promotion so that the new end is the last completed epoch
- This means that the `destroyPromotion` grace period will now start after the promotion ends, whether ended manually or through time.

Let's also remove the `setGracePeriod` function.  I don't think any creator will extend the grace period.  The less code the better!  Also nothing else is editable on the promotion so better to stay consistent IMO.

PierrickGT commented:
> `destroyPromotion` should play nicely with the existing `cancelPromotion`.
> 
> To better define them, and to ensure they have good interplay, let's do this:
> 
> * Rename `cancelPromotion` to `endPromotion`: because it allows the creator to prematurely end the promotion.
> * When the creator calls `endPromotion`, the code actually adjusts the promotion so that the new end is the last completed epoch
> * This means that the `destroyPromotion` grace period will now start after the promotion ends, whether ended manually or through time.
> 
> Let's also remove the `setGracePeriod` function. I don't think any creator will extend the grace period. The less code the better! Also nothing else is editable on the promotion so better to stay consistent IMO.

I've removed the `setGracePeriod` function.
The flow you describe was actually already implemented when you reviewed.
When calling `endPromotion` we adjust the total number of epochs with [`_promotions[_promotionId].numberOfEpochs = uint8(_getCurrentEpochId(_promotion));`](https://github.com/pooltogether/v4-periphery/blob/296a2d857d80e923dc3176dfef8e0291c0bc537b/contracts/TwabRewards.sol#L164).
This way, users can still claim rewards from previous epochs, including the one currently running.
Since we've removed `setGracePeriod`, [`_gracePeriod`](https://github.com/pooltogether/v4-periphery/blob/296a2d857d80e923dc3176dfef8e0291c0bc537b/contracts/TwabRewards.sol#L31) is now a private variable.
When calling [`destroyPromotion`](https://github.com/pooltogether/v4-periphery/blob/296a2d857d80e923dc3176dfef8e0291c0bc537b/contracts/TwabRewards.sol#L175) we check that the promotion is inactive with the require statement [(_getPromotionEndTimestamp(_promotion) + _gracePeriod) < block.timestamp](https://github.com/pooltogether/v4-periphery/blob/296a2d857d80e923dc3176dfef8e0291c0bc537b/contracts/TwabRewards.sol#L182).
When ending the promotion, we set the new total number of promotions, so `_getPromotionEndTimestamp(_promotion)` will return the end timestamp of the promotion with the formula [`_promotion.startTimestamp + (_promotion.epochDuration * _promotion.numberOfEpochs)`](https://github.com/pooltogether/v4-periphery/blob/296a2d857d80e923dc3176dfef8e0291c0bc537b/contracts/TwabRewards.sol#L356) and by adding the `_gracePeriod`, we ensure that the promotion can only be destroyed at least 60 days after the promotion was ended.

asselstine commented on contracts/TwabRewards.sol:
```diff
@@ -335,7 +393,9 @@ contract TwabRewards is ITwabRewards {
         if (block.timestamp > _promotion.startTimestamp) {
             unchecked {
                 // elapsedTimestamp / epochDurationTimestamp
```
✂️

PierrickGT replied:
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/ac29ad436f9460f7eef04eb838a2306387498b24

asselstine commented on contracts/TwabRewards.sol:
```diff
@@ -137,19 +156,47 @@ contract TwabRewards is ITwabRewards {
+
+        require(
+            (_getPromotionEndTimestamp(_promotion) + _gracePeriod) < block.timestamp,
+            "TwabRewards/promotion-active"
```
This is not accurate; the promotion may have ended but the grace period has not elapsed.

I think you should compare the **greater of** the promotion end or the created timestamp

i.e.

```
_requirePromotionInactive(_promotion)
graceEnd = (promotionEnd < createdAt ? createdAt : promotionEnd) + gracePeriod
require(block.timestamp >= graceEnd, "TwabRewards/grace-period-active")
```

That should give you the same information / errors

PierrickGT replied:
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/f2fce3efe5158e62a50a68ff5c4e2fa1d9b0cf47

asselstine commented on contracts/TwabRewards.sol:
```diff
@@ -168,20 +215,21 @@ contract TwabRewards is ITwabRewards {
             "TwabRewards/epochs-over-limit"
         );
 
         uint8 _extendedNumberOfEpochs = _currentNumberOfEpochs + _numberOfEpochs;
```
No need for this variable- it's only used once

PierrickGT replied:
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/ac29ad436f9460f7eef04eb838a2306387498b24

asselstine commented on contracts/TwabRewards.sol:
```diff
@@ -168,20 +215,21 @@ contract TwabRewards is ITwabRewards {
 
         unchecked {
-            _extendedAmount = _numberOfEpochs * _promotion.tokensPerEpoch;
+            _amount = _numberOfEpochs * _promotion.tokensPerEpoch;
```
Why is this unchecked?

PierrickGT replied:
`_numberOfEpochs` is a uint8 and `_promotion.tokensPerEpoch` is uint256, so it can't overflow since _amount is a uint256.

asselstine replied:
Same as above

PierrickGT replied:
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/0c7a7d0876b19120f26d68f58b5b6be4a092225e

asselstine commented on contracts/TwabRewards.sol:
```diff
@@ -108,21 +125,23 @@ contract TwabRewards is ITwabRewards {
+            _amount = _tokensPerEpoch * _numberOfEpochs;
+        }
+
         _promotions[_nextPromotionId] = Promotion(
```
Would be safer to use named variables here.  Just good practice!

i.e.

```
 = Promotion({
startTimestamp: _startTimestamp
// etc
})
```

PierrickGT replied:
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/86e130f68e5beac4fa7078fa560d475cfb679582

PierrickGT commented on contracts/TwabRewards.sol:
```diff
@@ -108,21 +125,23 @@ contract TwabRewards is ITwabRewards {
+        uint256 _amount;
+
+        unchecked {
+            _amount = _tokensPerEpoch * _numberOfEpochs;
```
`_tokensPerEpoch` is a `uint256` and `_numberOfEpochs` is `uint8`, so it can't overflow since `_amount` is a `uint256`.

asselstine replied:
That still doesn't make sense to me, because `2^256 * 2^8 = 2^264`.  Since 2^264 doesn't fit into 2^256 it would overflow.

PierrickGT replied:
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/0c7a7d0876b19120f26d68f58b5b6be4a092225e

PierrickGT commented on contracts/TwabRewards.sol:
```diff
@@ -27,6 +27,9 @@ contract TwabRewards is ITwabRewards {
     ITicket public ticket;
 
+    /// @notice Period during which the promotion owner can't destroy an inactive promotion.
+    uint32 internal _gracePeriod = 5184000; // 60 days in seconds
```
Done: https://github.com/pooltogether/v4-periphery/pull/48/commits/51b0b2c30062e97be0d17748bb03b15ec399d47a
